lgc/patch/Continufy.cpp

```diff
@@ -0,0 +1,196 @@
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ *
+ **********************************************************************************************************************/
+/**
+ ***********************************************************************************************************************
+ * @file  Continufy.cpp
+ * @brief LLPC source file: contains implementation of class lgc::Continufy.
+ * This pass translate indirect call into cps.await call, which will be lowered into continuation call.
+ ***********************************************************************************************************************
+ */
+#include "lgc/patch/Continufy.h"
+#include "lgccps/LgcCpsDialect.h"
+#include "lgcrt/LgcRtDialect.h"
+#include "lgc/Builder.h"
+#include "lgc/LgcDialect.h"
+#include "lgc/patch/Patch.h"
+#include "lgc/state/PalMetadata.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "lgc-continufy"
+
+using namespace llvm;
+using namespace lgc;
+
+namespace lgc {
+
+static Function *insertCpsArguments(Function &fn) {
+  // convert rtstage to levels
+  // Mutate function arguments, add ({} %state, %rcr).
+  LLVMContext &context = fn.getContext();
+  SmallVector<Type *> argTys = {StructType::get(context, {}), IntegerType::get(context, 32)};
+  auto *fnTy = fn.getFunctionType();
+  argTys.append(fnTy->params().begin(), fnTy->params().end());
+
+  auto *newFn = mutateFunctionArguments(fn, Type::getVoidTy(context), argTys, fn.getAttributes());
+
+  fn.replaceAllUsesWith(newFn);
+  for (unsigned idx = 0; idx < fn.arg_size(); idx++) {
+    Value *oldArg = fn.getArg(idx);
+    Value *newArg = newFn->getArg(idx + 2);
+    newArg->setName(oldArg->getName());
+    oldArg->replaceAllUsesWith(newArg);
+  }
+  newFn->getArg(0)->setName("state");
+  newFn->getArg(1)->setName("rcr");
+  return newFn;
+}
+
+/// Return the levels of the caller ray-tracing stages, which will be executed after the function return.
+static unsigned getReturnedLevels(int32_t stage) {
+  // Traversal will return to RGS.
+  if (stage == -1)
+    return 1;
+  rt::RayTracingShaderStage rtStage = static_cast<rt::RayTracingShaderStage>(stage);
+  switch (rtStage) {
+  case rt::RayTracingShaderStage::ShaderStageRayGeneration:
+    llvm_unreachable("Raygen shader should arrive here.\n");
+  case rt::RayTracingShaderStage::ShaderStageClosestHit:
+  case rt::RayTracingShaderStage::ShaderStageMiss:
+    // Traversal | RGS
+    return (1 << 2 | 1);
+  case rt::RayTracingShaderStage::ShaderStageCallable:
+    // CHS/Miss/Callable | Traversal | RGS
+    return (1 << 2 | 1 << 1 | 1);
+  case rt::RayTracingShaderStage::ShaderStageAnyHit:
+    // IS | Traversal | RGS
+    return (1 << 4 | 1 << 2 | 1);
+  default:
+    llvm_unreachable("Unknown raytracing shader type.\n");
+  }
+}
+
+/// Return level of the ray-tracing stage.
+static unsigned getLevelOfRtStage(int32_t stage) {
```

Comment:
Nit: Maybe return `uint32_t` instead of `unsigned`, given you also specify a bit width on the argument?